crypto/bn/bn_dsa.c

```diff
@@ -0,0 +1,137 @@
+/*
+ * Copyright 2019 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+#include <stdio.h>
+/* The line above prevents "ISO C forbids an empty translation unit" warning */
+#ifndef FIPS_MODE
+# include <string.h>
+# include <openssl/evp.h>
+# include <openssl/rand_drbg.h>
+# include "internal/cryptlib.h"
+# include "internal/bn_int.h"
+# include "../rand/rand_lcl.h"
+
+typedef struct drbg_ctx_st
+{
+    const unsigned char* x;
+    const unsigned char* h;
+    size_t rlen;
+} DRBG_CTX;
+
+static int app_data_index = 0;
+static size_t get_entropy(RAND_DRBG *drbg, unsigned char **pout,
+                          int entropy, size_t min_len, size_t max_len,
+                          int prediction_resistance)
+{
+    DRBG_CTX *drbg_ctx = (DRBG_CTX *)RAND_DRBG_get_ex_data(drbg, app_data_index);
+
+    *pout = (unsigned char *)drbg_ctx->x;
+    return drbg_ctx->rlen;
+}
+
+static size_t get_nonce(RAND_DRBG *drbg, unsigned char **pout,
+                        int entropy, size_t min_len, size_t max_len)
+{
+    DRBG_CTX *drbg_ctx = (DRBG_CTX *)RAND_DRBG_get_ex_data(drbg, app_data_index);
+
+    *pout = (unsigned char *)drbg_ctx->h;
+    return drbg_ctx->rlen;
+}
+
+static int bits2int(BIGNUM *out, int qlen,
+    const unsigned char *message, size_t message_len)
+{
+
+    if (BN_bin2bn(message, (int)message_len, out) == NULL)
+        return 0;
+    if ((int)message_len * 8 > qlen)
+        return BN_rshift(out, out, (int)message_len * 8 - qlen);
+    return 1;
+}
+static int int2octects(unsigned char *out, const BIGNUM  *num, int rlen)
+{
+    return BN_bn2binpad(num, out, rlen);
+}
+static int bits2octects(unsigned char *out, const BIGNUM  *range,
+    const unsigned char *message, size_t message_len, BN_CTX  *ctx)
+{
+    BIGNUM *num = BN_new();
+    if (
+        !bits2int(num, BN_num_bits(range), message, message_len) ||
+        !BN_mod(num, num, range, ctx) ||
+        !BN_bn2binpad(num, out, BN_num_bytes(range)))
+    {
+        BN_free(num);
+        return 0;
+    }
+    BN_clear_free(num);
+    return 1;
+}
+int bn_generate_dsa_deterministic_nonce(BIGNUM *out, const BIGNUM *range,
+    const BIGNUM *priv, const unsigned char *message,
+    size_t message_len, int hash_type, BN_CTX *ctx)
+{
+    int ret = 0;
+
+    RAND_DRBG *drbg = NULL;
+    DRBG_CTX drbg_ctx;
+    unsigned int drbg_flags = RAND_DRBG_FLAG_HMAC;
+    int rlen = 0, qlen = 0;
+    unsigned char *x = NULL, *h = NULL, *T = NULL;
+
+    app_data_index = RAND_DRBG_get_ex_new_index(0L, NULL, NULL, NULL, NULL);
```

Comment:
@slontis I think it is thread-unsafe here... But I failed to solve this problem.

Reply:
@mattcaswell  is there some complex way of doing this using OPENSSL_CTX (rand_crng_test does something).

Reply:
This function should only be called once at library initialisation and the index saved in the global variable.  All DRBG's can share the same value.  The first argument should be `CRYPTO_EX_INDEX_DSA`.

I'm not sure that this is the best approach for what is a transient DRBG.

Reply:
@paulidale 
Whether change this line to below is enough?
```
app_data_index = RAND_DRBG_get_ex_new_index(CRYPTO_EX_INDEX_DSA, NULL, NULL, NULL, NULL);
```
or I have to judge whether it is initialized like
```
if(app_data_index < 0) /* uninitialized */
    app_data_index = RAND_DRBG_get_ex_new_index(CRYPTO_EX_INDEX_DSA, NULL, NULL, NULL, NULL);
```

Reply:
Yeah this is bad.
We will need to look into another way of doing this.
It can be done through the OSSL_CTX but it is very ugly.
This may require another PR to fix this up.. @paulidale did you look at this?

Reply:
Another option is to add a new built-in index to openssl.  deterministic DSA is likely to be common, and it's only a define (and update the exdata manpage!)

Reply:
@slontis I did an initial stab at a test-DRBG that fed back what it was given -- this could be setup as the parent of the HMAC DRBG to avoid some of the nastiness.

Then I got distracted.

Reply:
```
if(app_data_index < 0) /* uninitialized */
    app_data_index = RAND_DRBG_get_ex_new_index(CRYPTO_EX_INDEX_DSA, NULL, NULL, NULL, NULL);
```
This won't work in a multi-threaded program.  Have a look at the `RUN_ONCE` macro and `DEFINE_RUN_ONCE_STATIC`.

@richsalz's suggestion might be cleaner.

Reply:
@paulidale @richsalz 
I don't know how to define a built in index even of I have read the man page...
Is there any example that have done this before?